<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Gene set enrichment analyses – Bioinformatics 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/phylogenomics.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a34d670291f06f286357e447776a572a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/gene_set_enrichment.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gene set enrichment analyses</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Bioinformatics 2</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Phylogenomics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/gene_set_enrichment.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gene set enrichment analyses</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#goals-of-this-lesson" id="toc-goals-of-this-lesson" class="nav-link active" data-scroll-target="#goals-of-this-lesson"><span class="header-section-number">2.1</span> Goals of this lesson</a></li>
  <li><a href="#enrichment-for-functional-terms" id="toc-enrichment-for-functional-terms" class="nav-link" data-scroll-target="#enrichment-for-functional-terms"><span class="header-section-number">2.2</span> Enrichment for functional terms</a></li>
  <li><a href="#enrichment-for-sequence-motifs" id="toc-enrichment-for-sequence-motifs" class="nav-link" data-scroll-target="#enrichment-for-sequence-motifs"><span class="header-section-number">2.3</span> Enrichment for sequence motifs</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Gene set enrichment analyses</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Gene set enrichment analyses are commonly used to identify overrepresented terms (mostly functional categories, but also sequence motifs and other sets) among a particular set of terms. For instance, researchers typically want to find if a list of differentially expressed genes (DEGs) is overrepresented in genes related to a specific pathway. To do that, we use statistical methods to calculate the probability of finding the frequency of terms we found by chance (under the null hypothesis). Although simple, (correctly) performing enrichment analyses requires special attention to avoid common mistakes and pitfalls.</p>
<section id="goals-of-this-lesson" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="goals-of-this-lesson"><span class="header-section-number">2.1</span> Goals of this lesson</h2>
<p>In this lesson, you will learn to:</p>
<ul>
<li>perform enrichment analyses for functional terms among differentially expressed genes;</li>
<li>perform enrichment analyses for sequence motifs among a set of promoter sequences.</li>
</ul>
<p>We will start by loading the required packages.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(universalmotif)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># for reproducibility</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="enrichment-for-functional-terms" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="enrichment-for-functional-terms"><span class="header-section-number">2.2</span> Enrichment for functional terms</h2>
<p>To start, we will load an RNA-seq data set stored as a <code>SummarizedExperiment</code> object in <code>data/GSE96870_se.rds</code>. Data were obtained from <span class="citation" data-cites="blackmore2017influenza">Blackmore et al. (<a href="#ref-blackmore2017influenza" role="doc-biblioref">2017</a>)</span> and modified to keep only a subset of the samples. In this experiment, researchers were trying to investigate how gene expression changes in male and female mice 4 and 8 days after infection with Influenza A. Samples were extracted from spinal cord and cerebellum tissues, but here we will only cerebellum samples only.</p>
<p>Let’s load the <code>SummarizedExperiment</code> object and take a look at it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"GSE96870_se.rds"</span>))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect data</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Entire object</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 41786 22 
metadata(0):
assays(1): counts
rownames(41786): Xkr4 LOC105243853 ... TrnT TrnP
rowData names(3): ENTREZID product gbkey
colnames(22): GSM2545337 GSM2545338 ... GSM2545346 GSM2545347
colData names(12): title geo_accession ... Label Group</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Count matrix</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             GSM2545337 GSM2545338 GSM2545348 GSM2545353 GSM2545343 GSM2545349
Xkr4               2410       2159       2275       1910       2235       1881
LOC105243853          0          1          1          0          3          0
LOC105242387        121        110        161        214        130        154
LOC105242467          5          5          2          1          2          4
Rp1                   2          0          3          1          1          6
Sox17               239        218        302        322        296        286
             GSM2545354 GSM2545339 GSM2545344 GSM2545352 GSM2545362 GSM2545340
Xkr4               1771       1980       1779       1890       2315       1977
LOC105243853          0          4          3          1          1          0
LOC105242387        124        120        131        272        189        172
LOC105242467          4          5          2          3          2          2
Rp1                   3          3          1          5          3          2
Sox17               273        220        233        267        197        261
             GSM2545345 GSM2545350 GSM2545363 GSM2545336 GSM2545342 GSM2545351
Xkr4               1528       2584       1645       1891       1757       1837
LOC105243853          0          0          0          0          1          1
LOC105242387        160        124        223        204        177        221
LOC105242467          2          7          1         12          3          1
Rp1                   2          5          1          2          3          3
Sox17               271        325        310        251        179        201
             GSM2545380 GSM2545341 GSM2545346 GSM2545347
Xkr4               1723       1945       1644       1585
LOC105243853          1          0          1          3
LOC105242387        251        173        180        176
LOC105242467          4          6          1          2
Rp1                   0          1          2          2
Sox17               246        232        205        230</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Sample metadata</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 22 rows and 12 columns
                     title geo_accession     organism         age         sex
               &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt; &lt;character&gt;
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus     8 weeks      Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus     8 weeks      Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus     8 weeks      Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus     8 weeks      Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus     8 weeks        Male
...                    ...           ...          ...         ...         ...
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus     8 weeks      Female
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus     8 weeks      Female
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus     8 weeks        Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus     8 weeks        Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus     8 weeks        Male
             infection      strain        time      tissue     mouse
           &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;integer&gt;
GSM2545337 NonInfected     C57BL/6        Day0  Cerebellum         9
GSM2545338 NonInfected     C57BL/6        Day0  Cerebellum        10
GSM2545348 NonInfected     C57BL/6        Day0  Cerebellum         8
GSM2545353 NonInfected     C57BL/6        Day0  Cerebellum         4
GSM2545343 NonInfected     C57BL/6        Day0  Cerebellum        11
...                ...         ...         ...         ...       ...
GSM2545351  InfluenzaA     C57BL/6        Day8  Cerebellum        16
GSM2545380  InfluenzaA     C57BL/6        Day8  Cerebellum        19
GSM2545341  InfluenzaA     C57BL/6        Day8  Cerebellum         6
GSM2545346  InfluenzaA     C57BL/6        Day8  Cerebellum        23
GSM2545347  InfluenzaA     C57BL/6        Day8  Cerebellum        24
                    Label       Group
                 &lt;factor&gt;    &lt;factor&gt;
GSM2545337 Female_Day0_9  Female_Day0
GSM2545338 Female_Day0_10 Female_Day0
GSM2545348 Female_Day0_8  Female_Day0
GSM2545353 Female_Day0_4  Female_Day0
GSM2545343 Male_Day0_11   Male_Day0  
...                   ...         ...
GSM2545351 Female_Day8_16 Female_Day8
GSM2545380 Female_Day8_19 Female_Day8
GSM2545341 Male_Day8_6    Male_Day8  
GSM2545346 Male_Day8_23   Male_Day8  
GSM2545347 Male_Day8_24   Male_Day8  </code></pre>
</div>
</div>
<p>Now, we will identify genes that are differentially expressed on day 8 after infection compared to day 0 (control).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep only protein-coding genes</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>gbkey <span class="sc">==</span> <span class="st">"mRNA"</span>]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get gene-wise differential expression test statistics (design: sex + time)</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSet</span>(se, <span class="at">design =</span> <span class="sc">~</span> sex <span class="sc">+</span> time)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain results from the desired contrast: day 8 vs day 0</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span>))</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get differentially expressed genes (DEGs) in a character vector</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>deg <span class="ot">&lt;-</span> res <span class="sc">|&gt;</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(padj <span class="sc">&lt;</span><span class="fl">0.05</span>, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;</span> <span class="fu">log2</span>(<span class="fl">1.5</span>)) <span class="sc">|&gt;</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"gene"</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(gene)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(deg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "3110035E14Rik" "Sgk3"          "Kcnb2"         "Sbspon"       
[5] "Gsta3"         "Lman2l"       </code></pre>
</div>
</div>
<p>Next, we will try find out if this set of DEGs is overrepresented in genes associated with a particular Gene Ontology (GO) term. For that, we will GO annotation in the <code>OrgDb</code> package for mouse. Bioconductor provides so-called <code>OrgDb</code> packages with annotation data for a bunch of model organisms, and many packages (including <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler">clusterProfiler</a></em>, which we will use to perform overrepresentation analysis) can use OrgDb packages to easily and automatically retrieve data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform ORA for GO terms</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ora_go <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene =</span> deg, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">keyType =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ont =</span> <span class="st">"BP"</span>, </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">OrgDb =</span> org.Mm.eg.db</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore results as a data frame</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(ora_go) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   ID                Description GeneRatio   BgRatio
GO:0050900 GO:0050900        leukocyte migration    51/968 402/28905
GO:0006935 GO:0006935                 chemotaxis    52/968 465/28905
GO:0042330 GO:0042330                      taxis    52/968 467/28905
GO:0030595 GO:0030595       leukocyte chemotaxis    36/968 242/28905
GO:0071674 GO:0071674 mononuclear cell migration    32/968 203/28905
GO:0060326 GO:0060326            cell chemotaxis    41/968 334/28905
                 pvalue     p.adjust       qvalue
GO:0050900 3.536123e-16 1.795643e-12 1.304643e-12
GO:0006935 3.336947e-14 6.712316e-11 4.876903e-11
GO:0042330 3.965528e-14 6.712316e-11 4.876903e-11
GO:0030595 6.591196e-14 8.367524e-11 6.079511e-11
GO:0071674 3.208410e-13 3.258461e-10 2.367469e-10
GO:0060326 8.641191e-13 7.313328e-10 5.313575e-10
                                                                                                                                                                                                                                                                                                               geneID
GO:0050900              Tnfsf18/Sell/Slamf9/Fut7/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Ascl2/Calr/Ccl17/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/Il33/Ch25h
GO:0006935 Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/Stx3/Ch25h/Plxnb3/Nox1
GO:0042330 Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Ntf3/Trpm4/Hsd3b7/Itgam/Adam8/Lsp1/Calr/Ccl17/Robo3/Cmtm7/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Tubb2b/Ccl28/Lgals3/Cmtm5/Ptk2b/Nr4a1/Casr/Retnlg/Fpr2/Dusp1/Ager/Stx3/Ch25h/Plxnb3/Nox1
GO:0030595                                                                                              Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Retnlg/Fpr2/Dusp1/Ch25h
GO:0071674                                                                                                                      Tnfsf18/Slamf9/Fut7/Itga4/Mdk/Grem1/Il12a/Nbl1/Padi2/Alox5/Trpm4/Hsd3b7/Adam8/Ascl2/Calr/Ccl17/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Lgals3/Ptk2b/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/Ch25h
GO:0060326                                                                Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Fpr2/Dusp1/Ch25h/Plxnb3/Nox1
           Count
GO:0050900    51
GO:0006935    52
GO:0042330    52
GO:0030595    36
GO:0071674    32
GO:0060326    41</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Overrepresentation analyses with non-model organisms
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you’re working with a non-model organism for which an <code>OrgDb</code> package is not available, you can create a 2-column data frame with functional annotation in the first column and gene IDs in the second column. Then, you can pass this annotation data frame to the <code>enricher()</code> function in <em><a href="https://bioconductor.org/packages/3.19/clusterProfiler">clusterProfiler</a></em>, which is more flexible and does not expect <code>OrgDb</code> objects as input.</p>
</div>
</div>
<p>As you can see, the <code>enrichGO()</code> function returned GO terms that are significantly overrepresented in our set of DEGs. However, there is an important issue with the way we performed the enrichment analysis: <code>enrichGO()</code> used all genes in the mouse genome as background (or ‘universe’). This does not make sense, because we didn’t use all genes for our differential expression analysis (e.g., we removed non-protein-coding genes). Choosing the right universe is <strong>very important</strong>, because a larger universe leads to more false positives (see an example <a href="https://carpentries-incubator.github.io/bioc-rnaseq/07-gene-set-analysis.html#choose-a-proper-universe">here</a>).</p>
<p>Now, let’s perform the enrichment analysis again, but this time restricting our universe to only those genes in the expression matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform ORA for GO terms with restricted universe</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>ora_go2 <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">gene =</span> deg, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">keyType =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ont =</span> <span class="st">"BP"</span>, </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">universe =</span> <span class="fu">rownames</span>(se),</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">OrgDb =</span> org.Mm.eg.db</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore results as a data frame</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(ora_go2) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   ID
GO:0050900 GO:0050900
GO:0071674 GO:0071674
GO:0030595 GO:0030595
GO:0002523 GO:0002523
GO:0035456 GO:0035456
GO:0002460 GO:0002460
                                                                                                                         Description
GO:0050900                                                                                                       leukocyte migration
GO:0071674                                                                                                mononuclear cell migration
GO:0030595                                                                                                      leukocyte chemotaxis
GO:0002523                                                                     leukocyte migration involved in inflammatory response
GO:0035456                                                                                               response to interferon-beta
GO:0002460 adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains
           GeneRatio   BgRatio       pvalue     p.adjust       qvalue
GO:0050900    51/968 372/16569 1.116320e-08 3.694287e-05 3.332087e-05
GO:0071674    32/968 180/16569 1.468901e-08 3.694287e-05 3.332087e-05
GO:0030595    36/968 222/16569 2.301099e-08 3.858177e-05 3.479908e-05
GO:0002523    11/968  25/16569 5.344683e-08 5.632460e-05 5.080235e-05
GO:0035456    16/968  55/16569 5.598867e-08 5.632460e-05 5.080235e-05
GO:0002460    44/968 320/16569 1.044964e-07 8.760284e-05 7.901396e-05
                                                                                                                                                                                                                                                                                                  geneID
GO:0050900 Tnfsf18/Sell/Slamf9/Fut7/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Ascl2/Calr/Ccl17/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/Il33/Ch25h
GO:0071674                                                                                                         Tnfsf18/Slamf9/Fut7/Itga4/Mdk/Grem1/Il12a/Nbl1/Padi2/Alox5/Trpm4/Hsd3b7/Adam8/Ascl2/Calr/Ccl17/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Lgals3/Ptk2b/Apod/Retnlg/Plg/Fpr2/Dusp1/Ager/Ch25h
GO:0030595                                                                                 Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Retnlg/Fpr2/Dusp1/Ch25h
GO:0002523                                                                                                                                                                                                                                   Fut7/Mdk/S100a8/S100a9/Ppbp/Ptn/Alox5/Itgam/Adam8/Ccl2/Aoc3
GO:0035456                                                                                                                                                                                        Aim2/Ifi204/Gbp6/Oas1c/Ifitm6/Bst2/Irgm1/Tgtp1/Tgtp2/Ifi47/Igtp/Irgm2/Ifitm7/F830016B08Rik/Iigp1/Ifit1
GO:0002460                              Il18rap/Pla2g4a/Tnfsf18/Fcgr4/Cd46/Fut7/C8g/Ada/Il12a/Rorc/Masp2/Fosl2/Pf4/Clec7a/Pirb/Trpm4/Irf7/Ascl2/Tnfsf13b/Il12rb1/Parp3/Enpp1/Aire/Icosl/Ccr7/Gzmb/Csf2rb/Il2rb/Tnfrsf13c/Emp2/Klhl6/Tfrc/H2-K1/Tap2/Ager/C4b/H2-Q1/H2-Q2/H2-Q6/H2-T23/Cd70/C3/Cd274/Il33
           Count
GO:0050900    51
GO:0071674    32
GO:0030595    36
GO:0002523    11
GO:0035456    16
GO:0002460    44</code></pre>
</div>
</div>
<p>Finally, you can also visually explore enrichment results as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(ora_go2, <span class="at">showCategory =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gene_set_enrichment_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Dotplot</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(ora_go2, <span class="at">showCategory =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gene_set_enrichment_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Although common, these visualization approaches are problematic because (i) they only show the top N (often 20) terms, when most of the time there are hundreds of enriched terms, and (ii) the GO database contains a lot of redundant terms, so many of the bars or dots in these plots actually represent the same thing. The best visualization technique for overrepresentation analysis is <strong>no visualization at all</strong>. Manually looking at the table and summarizing results in text is a much better thing to do.</p>
</section>
<section id="enrichment-for-sequence-motifs" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="enrichment-for-sequence-motifs"><span class="header-section-number">2.3</span> Enrichment for sequence motifs</h2>
<p>Enrichment analyses can also be used to find overrepresented sequence motifs among a set of sequences. For example, you may have a set of coexpressed genes and want to find out if a particular sequence motif is overrepresented in their promoter sequences (which would likely indicate transcription factor binding sites). There are many software tools for such kind of analysis, such as the MEME suite <span class="citation" data-cites="bailey2015meme">(<a href="#ref-bailey2015meme" role="doc-biblioref">Bailey et al. 2015</a>)</span> and the R/Bioconductor package universalmotif <span class="citation" data-cites="tremblay2024universalmotif">(<a href="#ref-tremblay2024universalmotif" role="doc-biblioref">Tremblay 2024</a>)</span>.</p>
<p>Here, we will use <em><a href="https://bioconductor.org/packages/3.19/universalmotif">universalmotif</a></em> to find out if an example motif (stored in the example data set <code>ArabidopsisMotif</code>) is significantly overrepresented in promoters of 50 genes in the genome of <em>Arabidopsis thaliana</em>. Let’s first load the example data and take a quick look at them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load example data: promoters and motifs of Arabidopsis thaliana</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"ArabidopsisMotif"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"ArabidopsisPromoters"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect data</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>ArabidopsisMotif</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
       Motif name:   YTTTYTTTTTYTTTY
         Alphabet:   DNA
             Type:   PPM
          Strands:   +-
         Total IC:   15.99
      Pseudocount:   1
        Consensus:   YTYTYTTYTTYTTTY
     Target sites:   617
          E-value:   2.5e-87

     Y    T    Y    T    Y    T    T    Y    T    T    Y    T    T    T    Y
A 0.01 0.00 0.00 0.00 0.00 0.06 0.00 0.01 0.00 0.00 0.02 0.00 0.00 0.00 0.00
C 0.30 0.17 0.31 0.01 0.54 0.02 0.24 0.25 0.22 0.04 0.39 0.21 0.16 0.18 0.43
G 0.16 0.05 0.03 0.01 0.00 0.02 0.11 0.00 0.04 0.05 0.03 0.01 0.02 0.00 0.11
T 0.53 0.78 0.66 0.98 0.45 0.90 0.66 0.74 0.74 0.91 0.55 0.77 0.83 0.82 0.46</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>ArabidopsisPromoters</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 50:
     width seq                                              names               
 [1]  1000 CATACAAGTAACAGAAAAACCGA...ACACTTTTGTCCGAAAACTAAG AT4G28150
 [2]  1000 TCGCTCATAGTTCTCCTAGATCG...AAATCCCAGTATTACCAACTTC AT1G19380
 [3]  1000 AATTGAAGGCTTTTACATCTTGG...TTGAAAGTCTTCAAGCAGACGC AT4G19520
 [4]  1000 TCGTAAAGTTAAAGGGAGACGTC...TTTAATAATCACATCTCTTACC AT1G03850
 [5]  1000 TACTTTTATGGATCATCATCAGA...TGTGTAATGACGATGAAAGATG AT5G01810
 ...   ... ...
[46]  1000 ACAGCAAAAACTGAGGTGCGAAC...CAGAGAGGACTGAGGAAATTCA AT5G22690
[47]  1000 GAATTGTGATACTATACTAATTG...GCCGTGCCCTTCATTTTCAGAA AT1G05670
[48]  1000 CATGCTTGGTTGGTTCTTCTTGT...AAACTTAGTGCCACTTAGACAC AT1G06160
[49]  1000 AGTTTATTGAATTAAGTTTTTAG...TAAGGAACGTTCAAGAACCAAA AT5G24660
[50]  1000 AGCGGTATATAAAAGTTAGTGCT...AGGATGTTGATGCATGAAAATC AT3G19200</code></pre>
</div>
</div>
<p>Now, to perform an enrichment analysis, we will use the <code>enrich_motifs()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform motif enrichment analysis</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">enrich_motifs</span>(</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">motifs =</span> ArabidopsisMotif, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequences =</span> ArabidopsisPromoters, </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">shuffle.k =</span> <span class="dv">3</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">threshold =</span> <span class="fl">0.001</span>, </span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">RC =</span> <span class="cn">TRUE</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 1 row and 15 columns
            motif   motif.i motif.consensus target.hits target.seq.hits
      &lt;character&gt; &lt;integer&gt;     &lt;character&gt;   &lt;integer&gt;       &lt;integer&gt;
1 YTTTYTTTTTYTTTY         1 YTYTYTTYTTYTTTY         244              50
  target.seq.count  bkg.hits bkg.seq.hits bkg.seq.count        Pval        Qval
         &lt;integer&gt; &lt;integer&gt;    &lt;integer&gt;     &lt;integer&gt;   &lt;numeric&gt;   &lt;numeric&gt;
1               50       135           45            50 1.23587e-08 1.23587e-08
         Eval pct.target.seq.hits pct.bkg.seq.hits target.enrichment
    &lt;numeric&gt;           &lt;numeric&gt;        &lt;numeric&gt;         &lt;numeric&gt;
1 2.47174e-08                 100               90           1.11111</code></pre>
</div>
</div>
<p>In principle, this analysis is very similar to what we’ve done for functional terms in the previous section: we are testing if input motifs are significantly overrepresented (or enriched) in a set of target sequences relative to a set of background sequences. Like in functional enrichment analyses, it’s important to carefully choose a suitable background (or universe). For instance, if you want to find overrepresented motifs in a set of coexpressed genes, you might want to use as background only promoters of expressed genes. When a background is not explicitly given, <code>enrich_motifs()</code> shuffles the input sequences to create a background.</p>
</section>
<section id="session-information" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-information">Session information</h2>
<p>This document was created under the following conditions:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 (2024-06-14)
 os       Ubuntu 22.04.4 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Arctic/Longyearbyen
 date     2025-03-12
 pandoc   3.2 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/x86_64/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package              * version  date (UTC) lib source
 abind                  1.4-5    2016-07-21 [1] CRAN (R 4.4.1)
 AnnotationDbi        * 1.66.0   2024-05-01 [1] Bioconductor 3.19 (R 4.4.1)
 ape                    5.8      2024-04-11 [1] CRAN (R 4.4.1)
 aplot                  0.2.3    2024-06-17 [1] CRAN (R 4.4.1)
 Biobase              * 2.64.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 BiocGenerics         * 0.50.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 BiocManager            1.30.23  2024-05-04 [1] CRAN (R 4.4.1)
 BiocParallel           1.38.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 BiocStyle              2.32.1   2024-06-16 [1] Bioconductor 3.19 (R 4.4.1)
 Biostrings             2.72.1   2024-06-02 [1] Bioconductor 3.19 (R 4.4.1)
 bit                    4.0.5    2022-11-15 [1] CRAN (R 4.4.1)
 bit64                  4.0.5    2020-08-30 [1] CRAN (R 4.4.1)
 blob                   1.2.4    2023-03-17 [1] CRAN (R 4.4.1)
 cachem                 1.1.0    2024-05-16 [1] CRAN (R 4.4.1)
 cli                    3.6.3    2024-06-21 [1] CRAN (R 4.4.1)
 clusterProfiler      * 4.12.1   2024-07-21 [1] Bioconductor 3.19 (R 4.4.1)
 codetools              0.2-20   2024-03-31 [1] CRAN (R 4.4.1)
 colorspace             2.1-0    2023-01-23 [1] CRAN (R 4.4.1)
 cowplot                1.1.3    2024-01-22 [1] CRAN (R 4.4.1)
 crayon                 1.5.3    2024-06-20 [1] CRAN (R 4.4.1)
 data.table             1.15.4   2024-03-30 [1] CRAN (R 4.4.1)
 DBI                    1.2.3    2024-06-02 [1] CRAN (R 4.4.1)
 DelayedArray           0.30.1   2024-05-07 [1] Bioconductor 3.19 (R 4.4.1)
 DESeq2               * 1.44.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 digest                 0.6.36   2024-06-23 [1] CRAN (R 4.4.1)
 DOSE                   3.30.2   2024-07-21 [1] Bioconductor 3.19 (R 4.4.1)
 dplyr                * 1.1.4    2023-11-17 [1] CRAN (R 4.4.1)
 enrichplot           * 1.24.2   2024-07-21 [1] Bioconductor 3.19 (R 4.4.1)
 evaluate               0.24.0   2024-06-10 [1] CRAN (R 4.4.1)
 fansi                  1.0.6    2023-12-08 [1] CRAN (R 4.4.1)
 farver                 2.1.2    2024-05-13 [1] CRAN (R 4.4.1)
 fastmap                1.2.0    2024-05-15 [1] CRAN (R 4.4.1)
 fastmatch              1.1-4    2023-08-18 [1] CRAN (R 4.4.1)
 fgsea                  1.30.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 forcats              * 1.0.0    2023-01-29 [1] CRAN (R 4.4.1)
 fs                     1.6.4    2024-04-25 [1] CRAN (R 4.4.1)
 generics               0.1.3    2022-07-05 [1] CRAN (R 4.4.1)
 GenomeInfoDb         * 1.40.1   2024-05-24 [1] Bioconductor 3.19 (R 4.4.1)
 GenomeInfoDbData       1.2.12   2024-07-24 [1] Bioconductor
 GenomicRanges        * 1.56.1   2024-06-12 [1] Bioconductor 3.19 (R 4.4.1)
 ggforce                0.4.2    2024-02-19 [1] CRAN (R 4.4.1)
 ggfun                  0.1.5    2024-05-28 [1] CRAN (R 4.4.1)
 ggplot2              * 3.5.1    2024-04-23 [1] CRAN (R 4.4.1)
 ggplotify              0.1.2    2023-08-09 [1] CRAN (R 4.4.1)
 ggraph                 2.2.1    2024-03-07 [1] CRAN (R 4.4.1)
 ggrepel                0.9.5    2024-01-10 [1] CRAN (R 4.4.1)
 ggtree                 3.12.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 glue                   1.7.0    2024-01-09 [1] CRAN (R 4.4.1)
 GO.db                  3.19.1   2024-07-24 [1] Bioconductor
 GOSemSim               2.30.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 graphlayouts           1.1.1    2024-03-09 [1] CRAN (R 4.4.1)
 gridExtra              2.3      2017-09-09 [1] CRAN (R 4.4.1)
 gridGraphics           0.5-1    2020-12-13 [1] CRAN (R 4.4.1)
 gson                   0.1.0    2023-03-07 [1] CRAN (R 4.4.1)
 gtable                 0.3.5    2024-04-22 [1] CRAN (R 4.4.1)
 HDO.db                 0.99.1   2024-07-24 [1] Bioconductor
 here                 * 1.0.1    2020-12-13 [1] CRAN (R 4.4.1)
 hms                    1.1.3    2023-03-21 [1] CRAN (R 4.4.1)
 htmltools              0.5.8.1  2024-04-04 [1] CRAN (R 4.4.1)
 htmlwidgets            1.6.4    2023-12-06 [1] CRAN (R 4.4.1)
 httr                   1.4.7    2023-08-15 [1] CRAN (R 4.4.1)
 igraph                 2.0.3    2024-03-13 [1] CRAN (R 4.4.1)
 IRanges              * 2.38.1   2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
 jsonlite               1.8.8    2023-12-04 [1] CRAN (R 4.4.1)
 KEGGREST               1.44.1   2024-06-19 [1] Bioconductor 3.19 (R 4.4.1)
 knitr                  1.48     2024-07-07 [1] CRAN (R 4.4.1)
 labeling               0.4.3    2023-08-29 [1] CRAN (R 4.4.1)
 lattice                0.22-6   2024-03-20 [1] CRAN (R 4.4.1)
 lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.4.1)
 lifecycle              1.0.4    2023-11-07 [1] CRAN (R 4.4.1)
 locfit                 1.5-9.10 2024-06-24 [1] CRAN (R 4.4.1)
 lubridate            * 1.9.3    2023-09-27 [1] CRAN (R 4.4.1)
 magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.4.1)
 MASS                   7.3-61   2024-06-13 [1] CRAN (R 4.4.1)
 Matrix                 1.7-0    2024-04-26 [1] CRAN (R 4.4.1)
 MatrixGenerics       * 1.16.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 matrixStats          * 1.3.0    2024-04-11 [1] CRAN (R 4.4.1)
 memoise                2.0.1    2021-11-26 [1] CRAN (R 4.4.1)
 munsell                0.5.1    2024-04-01 [1] CRAN (R 4.4.1)
 nlme                   3.1-165  2024-06-06 [1] CRAN (R 4.4.1)
 org.Mm.eg.db         * 3.19.1   2024-09-02 [1] Bioconductor
 patchwork              1.2.0    2024-01-08 [1] CRAN (R 4.4.1)
 pillar                 1.9.0    2023-03-22 [1] CRAN (R 4.4.1)
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.4.1)
 plyr                   1.8.9    2023-10-02 [1] CRAN (R 4.4.1)
 png                    0.1-8    2022-11-29 [1] CRAN (R 4.4.1)
 polyclip               1.10-7   2024-07-23 [1] CRAN (R 4.4.1)
 purrr                * 1.0.2    2023-08-10 [1] CRAN (R 4.4.1)
 qvalue                 2.36.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 R6                     2.5.1    2021-08-19 [1] CRAN (R 4.4.1)
 RColorBrewer           1.1-3    2022-04-03 [1] CRAN (R 4.4.1)
 Rcpp                   1.0.13   2024-07-17 [1] CRAN (R 4.4.1)
 readr                * 2.1.5    2024-01-10 [1] CRAN (R 4.4.1)
 reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.4.1)
 rlang                  1.1.4    2024-06-04 [1] CRAN (R 4.4.1)
 rmarkdown              2.27     2024-05-17 [1] CRAN (R 4.4.1)
 rprojroot              2.0.4    2023-11-05 [1] CRAN (R 4.4.1)
 RSQLite                2.3.7    2024-05-27 [1] CRAN (R 4.4.1)
 rstudioapi             0.16.0   2024-03-24 [1] CRAN (R 4.4.1)
 S4Arrays               1.4.1    2024-05-20 [1] Bioconductor 3.19 (R 4.4.1)
 S4Vectors            * 0.42.1   2024-07-03 [1] Bioconductor 3.19 (R 4.4.1)
 scales                 1.3.0    2023-11-28 [1] CRAN (R 4.4.1)
 scatterpie             0.2.3    2024-06-05 [1] CRAN (R 4.4.1)
 sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.4.1)
 shadowtext             0.1.4    2024-07-18 [1] CRAN (R 4.4.1)
 SparseArray            1.4.8    2024-05-24 [1] Bioconductor 3.19 (R 4.4.1)
 stringi                1.8.4    2024-05-06 [1] CRAN (R 4.4.1)
 stringr              * 1.5.1    2023-11-14 [1] CRAN (R 4.4.1)
 SummarizedExperiment * 1.34.0   2024-05-01 [1] Bioconductor 3.19 (R 4.4.1)
 tibble               * 3.2.1    2023-03-20 [1] CRAN (R 4.4.1)
 tidygraph              1.3.1    2024-01-30 [1] CRAN (R 4.4.1)
 tidyr                * 1.3.1    2024-01-24 [1] CRAN (R 4.4.1)
 tidyselect             1.2.1    2024-03-11 [1] CRAN (R 4.4.1)
 tidytree               0.4.6    2023-12-12 [1] CRAN (R 4.4.1)
 tidyverse            * 2.0.0    2023-02-22 [1] CRAN (R 4.4.1)
 timechange             0.3.0    2024-01-18 [1] CRAN (R 4.4.1)
 treeio                 1.28.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 tweenr                 2.0.3    2024-02-26 [1] CRAN (R 4.4.1)
 tzdb                   0.4.0    2023-05-12 [1] CRAN (R 4.4.1)
 UCSC.utils             1.0.0    2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 universalmotif       * 1.22.3   2024-10-02 [1] Bioconductor 3.19 (R 4.4.1)
 utf8                   1.2.4    2023-10-22 [1] CRAN (R 4.4.1)
 vctrs                  0.6.5    2023-12-01 [1] CRAN (R 4.4.1)
 viridis                0.6.5    2024-01-29 [1] CRAN (R 4.4.1)
 viridisLite            0.4.2    2023-05-02 [1] CRAN (R 4.4.1)
 withr                  3.0.0    2024-01-16 [1] CRAN (R 4.4.1)
 xfun                   0.46     2024-07-18 [1] CRAN (R 4.4.1)
 XVector                0.44.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)
 yaml                   2.3.9    2024-07-05 [1] CRAN (R 4.4.1)
 yulab.utils            0.1.4    2024-01-28 [1] CRAN (R 4.4.1)
 zlibbioc               1.50.0   2024-04-30 [1] Bioconductor 3.19 (R 4.4.1)

 [1] /home/faalm/R/x86_64-pc-linux-gnu-library/4.4
 [2] /usr/local/lib/R/site-library
 [3] /usr/lib/R/site-library
 [4] /usr/lib/R/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-bailey2015meme" class="csl-entry" role="listitem">
Bailey, Timothy L, James Johnson, Charles E Grant, and William S Noble. 2015. <span>“The MEME Suite.”</span> <em>Nucleic Acids Research</em> 43 (W1): W39–49.
</div>
<div id="ref-blackmore2017influenza" class="csl-entry" role="listitem">
Blackmore, Stephen, Jessica Hernandez, Michal Juda, Emily Ryder, Gregory G Freund, Rodney W Johnson, and Andrew J Steelman. 2017. <span>“Influenza Infection Triggers Disease in a Genetic Model of Experimental Autoimmune Encephalomyelitis.”</span> <em>Proceedings of the National Academy of Sciences</em> 114 (30): E6107–16.
</div>
<div id="ref-tremblay2024universalmotif" class="csl-entry" role="listitem">
Tremblay, Benjamin Jean-Marie. 2024. <span>“Universalmotif: An r Package for Biological Motif Analysis.”</span> <em>Journal of Open Source Software</em> 9 (100): 7012.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/phylogenomics.html" class="pagination-link" aria-label="Phylogenomics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Phylogenomics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>