<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bioinformatics 2 - 1&nbsp; Phylogenomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../index.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/phylogenomics.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Phylogenomics</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Bioinformatics 2</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/phylogenomics.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Phylogenomics</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#goals-of-this-lesson" id="toc-goals-of-this-lesson" class="nav-link active" data-scroll-target="#goals-of-this-lesson"><span class="header-section-number">1.1</span> Goals of this lesson</a></li>
  <li><a href="#data-acquisition" id="toc-data-acquisition" class="nav-link" data-scroll-target="#data-acquisition"><span class="header-section-number">1.2</span> Data acquisition</a></li>
  <li><a href="#obtaining-busco-genes" id="toc-obtaining-busco-genes" class="nav-link" data-scroll-target="#obtaining-busco-genes"><span class="header-section-number">1.3</span> Obtaining BUSCO genes</a></li>
  <li><a href="#multiple-sequence-alignment" id="toc-multiple-sequence-alignment" class="nav-link" data-scroll-target="#multiple-sequence-alignment"><span class="header-section-number">1.4</span> Multiple sequence alignment</a></li>
  <li><a href="#phylogeny-inference-concatenation-based-approaches" id="toc-phylogeny-inference-concatenation-based-approaches" class="nav-link" data-scroll-target="#phylogeny-inference-concatenation-based-approaches"><span class="header-section-number">1.5</span> Phylogeny inference: concatenation-based approaches</a></li>
  <li><a href="#phylogeny-inference-coalescent-based-approaches" id="toc-phylogeny-inference-coalescent-based-approaches" class="nav-link" data-scroll-target="#phylogeny-inference-coalescent-based-approaches"><span class="header-section-number">1.6</span> Phylogeny inference: coalescent-based approaches</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Phylogenomics</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Phylogenomics aims at reconstructing the evolutionary histories of organisms taking into account whole genomes or large fractions of genomes. In this lesson, you will learn how to infer a species tree from whole-genome sequence data using a standard phylogenomics pipeline.</p>
<section id="goals-of-this-lesson" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="goals-of-this-lesson"><span class="header-section-number">1.1</span> Goals of this lesson</h2>
<p>In this lesson, you will learn to:</p>
<ul>
<li>identify single-copy orthologs, typical markers for phylogenomic studies;</li>
<li>perform multiple sequence alignments;</li>
<li>infer phylogenetic trees using concatenation-based <span class="citation" data-cites="minh2020iq">(<a href="#ref-minh2020iq" role="doc-biblioref">Minh et al. 2020</a>)</span> and coalescent-based <span class="citation" data-cites="zhang2017astral">(<a href="#ref-zhang2017astral" role="doc-biblioref">Zhang, Sayyari, and Mirarab 2017</a>)</span> approaches;</li>
<li>read, explore, and visualize phylogenetic trees in R.</li>
</ul>
<p>We will start by loading the required packages.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cogeqc)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msa)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># for reproducibility</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="fl">1e8</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(<span class="st">"code"</span>, <span class="st">"utils.R"</span>)) <span class="co"># helper functions</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-acquisition" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="data-acquisition"><span class="header-section-number">1.2</span> Data acquisition</h2>
<p>In this lesson, we will use data from Ensembl <span class="citation" data-cites="harrison2024ensembl">(<a href="#ref-harrison2024ensembl" role="doc-biblioref">Harrison et al. 2024</a>)</span>, a database with genomic data for vertebrate species. However, since using all species on Ensembl would lead to much longer runtimes, we will only use species that are present on <em>Ensembl Pan Compara</em>, a cross-species comparative resource on Ensembl. We will start by retrieving species metadata.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get metadata for species on Pan Compara</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://ftp.ensembl.org/pub/release-113/species_EnsemblVertebrates.txt"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> <span class="cn">FALSE</span>, <span class="at">col_select =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(pan_compara <span class="sc">==</span> <span class="st">"Y"</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>metadata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 6
   `#name`              species division taxonomy_id assembly assembly_accession
   &lt;chr&gt;                &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;             
 1 Green anole          anolis… Ensembl…       28377 AnoCar2… GCA_000090745.2   
 2 C.savignyi           ciona_… Ensembl…       51511 CSAV 2.0 &lt;NA&gt;              
 3 Zebrafish            danio_… Ensembl…        7955 GRCz11   GCA_000002035.4   
 4 Chicken              gallus… Ensembl…        9031 bGalGal… GCA_016699485.1   
 5 Stickleback          gaster… Ensembl…      481459 GAculea… GCA_016920845.1   
 6 Human                homo_s… Ensembl…        9606 GRCh38.… GCA_000001405.29  
 7 Opossum              monode… Ensembl…       13616 ASM229v1 GCA_000002295.1   
 8 Mouse                mus_mu… Ensembl…       10090 GRCm39   GCA_000001635.9   
 9 Platypus             ornith… Ensembl…        9258 mOrnAna… GCA_004115215.2   
10 Chimpanzee           pan_tr… Ensembl…        9598 Pan_tro… GCA_000001515.5   
11 Sumatran orangutan   pongo_… Ensembl…        9601 Susie_P… GCA_002880775.3   
12 Tropical clawed frog xenopu… Ensembl…        8364 UCB_Xtr… GCA_000004195.4   </code></pre>
</div>
</div>
<p>Then, we will obtain whole-genome protein sequences (only one for each gene) for each species. To do that, we will define a helper function to:</p>
<ol type="1">
<li>Read FASTA files directly from Ensembl’s FTP repository as <code>AAStringSet</code> objects.</li>
<li>Filter the <code>AAStringSet</code> objects to keep only translated sequences of longest transcripts (i.e., one gene -&gt; one protein).</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define helper function to read proteomes from Ensembl</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>get_proteomes <span class="ot">&lt;-</span> <span class="cf">function</span>(species) {</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get URL to FTP directory</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    surl <span class="ot">&lt;-</span> <span class="fu">file.path</span>(</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"https://ftp.ensembl.org/pub/release-113/fasta"</span>, </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>        species, <span class="st">"pep/"</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get protein sequences of longest transcripts for each gene</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    seqs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(surl, <span class="cf">function</span>(x) {</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>        <span class="do">## Get links to FASTA files</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>        fname <span class="ot">&lt;-</span> XML<span class="sc">::</span><span class="fu">getHTMLLinks</span>(RCurl<span class="sc">::</span><span class="fu">getURL</span>(x, <span class="at">dirlistonly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>        fname <span class="ot">&lt;-</span> fname[<span class="fu">grep</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.fa</span><span class="sc">\\</span><span class="st">.gz$"</span>, fname)]</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>        fname <span class="ot">&lt;-</span> fname[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">"abinitio"</span>, fname)]</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>        seq_path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(x, fname)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>        <span class="do">## Read sequences as an `AAStringSet` object</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>        seq <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">readAAStringSet</span>(seq_path)</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(seq) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" .*"</span>, <span class="st">""</span>, <span class="fu">gsub</span>(<span class="st">".*gene:"</span>, <span class="st">""</span>, <span class="fu">names</span>(seq)))</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>        <span class="do">## Keep only translated sequence of longest transcript</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>        seq <span class="ot">&lt;-</span> seq[<span class="fu">order</span>(Biostrings<span class="sc">::</span><span class="fu">width</span>(seq), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>        seq <span class="ot">&lt;-</span> seq[<span class="sc">!</span><span class="fu">duplicated</span>(<span class="fu">names</span>(seq))]</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(seq)</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(seqs) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, species)</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(seqs)</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we will <em>actually</em> get the proteomes by executing the function we’ve just defined.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get proteomes for all species</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>proteomes <span class="ot">&lt;-</span> <span class="fu">get_proteomes</span>(metadata<span class="sc">$</span>species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="obtaining-busco-genes" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="obtaining-busco-genes"><span class="header-section-number">1.3</span> Obtaining BUSCO genes</h2>
<p>Now that we have whole-genome protein sequences for each species, we can run BUSCO to identify single-copy gene families in our proteomes. To do that, BUSCO relies on comparing our sequences with pre-built profile Hidden Markov Models (HMMs) for a particular clade. These profiles HMMs are available for a bunch of different clades (referred to as ‘lineage data sets’), and selecting the right lineage data set is critical to correctly identify single-copy gene families. To view a list of all lineage data sets available in BUSCO, you can run the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List lineage datasets</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>cogeqc<span class="sc">::</span><span class="fu">list_busco_datasets</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you can see, the lineage data sets in BUSCO have a hierarchical (i.e., tree-like) structure. For example, the lineage <em>vertebrata_odb10</em> contains two lineages, <em>actinopterygii_odb10</em> and <em>tetrapoda_odb10</em>. The lineage <em>tetrapoda_odb10</em>, in turn, contains the lineages <em>mammalia_odb10</em> and <em>sauropsida_odb10</em>. The best lineage data set is usually thought to be the one that includes all species in your data set in the most taxonomically restricted way. This is because the number of BUSCO genes decreases as we go from a broader clade (e.g.&nbsp;Vertebrata) to a stricter clade (e.g.&nbsp;Mammalia). For instance, if your data set contains only mammals, you could use lineages <em>metazoa_odb10</em>, <em>vertebrata_odb10</em>, <em>tetrapoda_odb10</em>, or <em>mammalia_odb10</em>. Let’s see the number of BUSCO genes for each of these clades.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot number of BUSCO genes for clades from metazoa to mammalia</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>clades <span class="ot">&lt;-</span> <span class="st">"metazoa|vertebrata|tetrapoda|mammalia"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>p_lineages <span class="ot">&lt;-</span> rvest<span class="sc">::</span><span class="fu">read_html</span>(<span class="st">"https://busco.ezlab.org/list_of_lineages.html"</span>) <span class="sc">|&gt;</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    rvest<span class="sc">::</span><span class="fu">html_table</span>() <span class="sc">|&gt;</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">lineage =</span> <span class="dv">1</span>, <span class="at">ngenes =</span> <span class="dv">2</span>) <span class="sc">|&gt;</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">str_detect</span>(lineage, clades)) <span class="sc">|&gt;</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="sc">-</span>ngenes) <span class="sc">|&gt;</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">lineage =</span> <span class="fu">as_factor</span>(lineage)) <span class="sc">|&gt;</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> ngenes, <span class="at">y =</span> lineage)) <span class="sc">+</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"deepskyblue4"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> ngenes), <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">1.2e4</span>), <span class="at">expand =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="fl">0.01</span>)) <span class="sc">+</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="st">"Number of BUSCO genes for each lineage"</span>,</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Number of genes"</span>, <span class="at">y =</span> <span class="st">"Lineage"</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>p_lineages</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="phylogenomics_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The plot shows that more restrict clades have more BUSCO genes. Thus, if you want to infer a phylogeny for some mammals, you should choose <em>mammalia_odb10</em> as your lineage data set, not broader clades that include other things beside mammals (e.g.&nbsp;<em>vertebrata_odb10</em> or <em>metazoa_odb10</em>).</p>
<p>The example data set in this lesson contains a diverse set of animals, most of which are vertebrates, but also with the tunicate <em>Ciona savignyi</em> (outgroup). Given this data set (vertebrates + a tunicate), the best lineage data set is <em>metazoa_odb10</em>.</p>
<p>To identify what BUSCO genes can be found in our data set, we can use the function <code>run_busco</code> from the R package <em><a href="https://bioconductor.org/packages/3.19/cogeqc">cogeqc</a></em> to run BUSCO in protein mode using <em>metazoa_odb10</em> as lineage data set. However, because running BUSCO takes several minutes, we will load a pre-computed set of BUSCO genes identified in this data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pre-computed BUSCO genes</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>busco_genes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"busco_seqs.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The file <code>busco_seqs.rds</code> was created with the code below, which does the following:</p>
<ol type="1">
<li>Run BUSCO in <code>protein</code> mode using <em>metazoa_odb10</em> as lineage data set.</li>
<li>Identify BUSCO genes that are shared by all species (i.e., species coverage of 100%);</li>
<li>Read sequences of conserved BUSCO genes as a list of <code>AAStringSet</code> objects, each containing 12 sequences (the BUSCO gene <em>i</em> in all 12 species).</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Show me the code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Export sequences to a directory</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>seq_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"seqs"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(seq_dir)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>e <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(proteomes), <span class="cf">function</span>(x) {</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    fn <span class="ot">&lt;-</span> <span class="fu">file.path</span>(seq_dir, <span class="fu">paste0</span>(<span class="fu">names</span>(proteomes)[x], <span class="st">".fasta"</span>))</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">writeXStringSet</span>(proteomes[[x]], <span class="at">filepath =</span> fn)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Run BUSCO in `protein` mode</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>busco_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"busco"</span>) <span class="co"># path to BUSCO output</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>data_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">tempdir</span>(), <span class="st">"datasets"</span>) <span class="co"># path to BUSCO datasets</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>cogeqc<span class="sc">::</span><span class="fu">run_busco</span>(</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">sequence =</span> seq_dir,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">outlabel =</span> <span class="st">"ensembl"</span>,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">mode =</span> <span class="st">"proteins"</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">lineage =</span> <span class="st">"metazoa_odb10"</span>,</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">outpath =</span> busco_output,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">download_path =</span> data_output,</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Read sequences of BUSCO genes</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>busco_seqdir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(busco_output, <span class="st">"ensembl"</span>)</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>busco_seqs <span class="ot">&lt;-</span> <span class="fu">read_busco_sequences</span>(busco_seqdir)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Save object to file</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(busco_seqs, <span class="at">compress =</span> <span class="st">"xz"</span>, <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"busco_seqs.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>This is what the object looks like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the first 3 BUSCO genes</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(busco_genes, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`102804at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]  1050 MALSLDREAYYRRIKRLYSNWQK...HHRSSHSRSPGRSSAPPKKKRK anolis.carolinensis
 [2]  1040 MSISLDKDAFFRRLKKIYSTWES...SDDDEGETMRRVGRIGMNWRRR ciona.savignyi
 [3]  1033 MAVSLDKDAFYRRIKRLYGNWKK...SNRKRKGRSSAPPPRSKKKRRH danio.rerio
 [4]   642 MEAVRKQKPELMAKITKNLGFAM...GRWGRWGRWGRWGHWGCWGQWG gallus.gallus
 [5]   938 MAVNLDKEAYYRRIKRFYGNWKK...EDSDEDYDSDTEDSSEILLFGH gasterosteus.acul...
 ...   ... ...
 [8]  1047 MAVTLDKDAYYRRVKRLYSNWRK...SSGRGSNRGSRHSSAPPKKKRK mus.musculus
 [9]  1047 MAVTLDKEAYYRRVKRLYSNWRK...SSGRGSGRGSRHSSAPPKKKRK ornithorhynchus.a...
[10]  1047 MAVTLDKDAYYRRVKRLYSNWRK...SSGRGSNRGSRHSSAPPKKKRK pan.troglodytes
[11]  1047 MAVTLDKDAYYRRVKRLYSNWRK...SSGRGSNRGSRHSSAPPKKKRK pongo.abelii
[12]  1035 MAVTLDKEAYYRRIKRFYGNWKK...QKGGKKRKVHAPAPNPSKKRKK xenopus.tropicalis

$`107151at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]   988 MAAAAASSSASGASSSSSSGPGG...NNWQSKIGEILVLVAAKIKNTF anolis.carolinensis
 [2]   739 NRSFHPQFSGYAGNLFSSGPYEK...KNWKLKTKELLPIISSKVPVAF ciona.savignyi
 [3]   974 MASSAAKQPPKNASKAPSGGAGG...LNWQKNIGEILVLVASKIKNSF danio.rerio
 [4]   999 MNKKKKPFLGMPAPLGYVPGLGR...VNWQKKIGEILVLVAAKLKNTF gallus.gallus
 [5]   973 MASTGIKQGPKLVSKSSAGGAGA...TVSRTSRSAVSRSVFPLKSITV gasterosteus.acul...
 ...   ... ...
 [8]   941 MNKKKKPFLGMPAPLGYVPGLGR...TNWQRKIGEILVLVAARIKNTF mus.musculus
 [9]   941 MNKKKKPFLGMPAPLGYVPGLGR...ENWQKKIGEILVLVAAKIKNAF ornithorhynchus.a...
[10]   941 MNKKKKPFLGMPAPLGYVPGLGR...ANWQKKIGDILRLVAGRIKNTF pan.troglodytes
[11]   964 MTLQNYYYYYYYYFFFFETKSCS...ANWQKKIGDILRLVAGRIKNTF pongo.abelii
[12]   972 MAASVSAGSGAGPSGSGNPTGTS...KNWQQKIGEILVLVAAQIKNAF xenopus.tropicalis

$`108764at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]   947 MSGFSPELIEYLEGKISFEEFER...RGGELAGSDRWCCLSFSEEQKP anolis.carolinensis
 [2]   771 SRLPRELRGLMGEANLCYARGDH...LIYRKSGNIDLANQTIHKYCVV ciona.savignyi
 [3]   918 MSELNDYLEGKITFEEFEKRREE...LIYQSSGNKDMARHIIYTYCTI danio.rerio
 [4]   927 MSGFSPELIDYLEGKISFEEFER...LIYHSSGNTRMAQKMLYTYAVV gallus.gallus
 [5]   944 MSGFSSELIDYLEGRITFEEFDQ...LIYQASGNMGMAHQLINTHCII gasterosteus.acul...
 ...   ... ...
 [8]   882 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNTAMAQKLLYTYCVI mus.musculus
 [9]   899 MSGFSPELIDYLEGKISFEEFER...LIYQNSGNLGLARRLLYTWGSV ornithorhynchus.a...
[10]   884 MSGFSPELIDYLEGKISSSSSNG...LIYQSSGNTGMAQTLLYTYCSI pan.troglodytes
[11]   885 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNTGMAQRLLYTYCSI pongo.abelii
[12]   880 MSGFSPELIDYLEGKISFEEFER...LIYQNSGNLEMARRLLHQYCVS xenopus.tropicalis</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check number of genes</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(busco_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 282</code></pre>
</div>
</div>
<p>As you can see, we found 282 BUSCO genes shared by all 12 species in our data set. These 282 genes represent our marker gene set for phylogeny inference.</p>
</section>
<section id="multiple-sequence-alignment" class="level2 page-columns page-full" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="multiple-sequence-alignment"><span class="header-section-number">1.4</span> Multiple sequence alignment</h2>
<p>Once we have our marker genes, we can perform a multiple sequence alignment (MSA) and use the MSA to infer a phylogeny.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform MSA for each BUSCO gene with MUSCLE</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>msas <span class="ot">&lt;-</span> <span class="fu">lapply</span>(busco_genes, msa, <span class="at">method =</span> <span class="st">"Muscle"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert MSAs to `AAStringSet` class</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>msas_aa <span class="ot">&lt;-</span> <span class="fu">lapply</span>(msas, as, <span class="st">"AAStringSet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a look at the first few MSAs to see what they look like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the MSAs for the first 3 genes</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(msas_aa, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`102804at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]  1123 MSISLDKDAFFRRLKKIYSTWES...-------------------RRR ciona.savignyi
 [2]  1123 -----------------------...WGQWG----------------- gallus.gallus
 [3]  1123 MAVTLDKEAYYRRIKRFYGNWK-...-----------PAPNPSKKRKK xenopus.tropicalis
 [4]  1123 MALSLDREAYYRRIKRLYSNWQ-...HHRSSHSRSPGRSSAPPKKKRK anolis.carolinensis
 [5]  1123 MAVTLDKDAYYRRVKRLYSNWR-...R---GSNRGSRHSSAPPKKKRK homo.sapiens
 ...   ... ...
 [8]  1123 MAVTLDKDAYYRRVKRLYSNWR-...R---GSNRGSRHSSAPPKKKRK mus.musculus
 [9]  1123 MAVTLDKEAYYRRVKRLYSNWR-...R---GSSRGSRHSSAPPKKKRK monodelphis.domes...
[10]  1123 MAVTLDKEAYYRRVKRLYSNWR-...R---GSGRGSRHSSAPPKKKRK ornithorhynchus.a...
[11]  1123 MAVSLDKDAFYRRIKRLYGNWK-...-------RSSAPPPRSKKKRRH danio.rerio
[12]  1123 MAVNLDKEAYYRRIKRFYGNWK-...---------------------- gasterosteus.acul...

$`107151at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]  1062 -----------------------...LLPII--SSKVPVAF------- ciona.savignyi
 [2]  1062 -----MASTGIKQGPKLVSKSSA...TVSRTSRSAVSRSVFPLKSITV gasterosteus.acul...
 [3]  1062 -----MASSAAKQPPKNASKAPS...ILVLV--ASKIKNSF------- danio.rerio
 [4]  1062 -------MAASVSAGSGAGPSGS...ILVLV--AAQIKNAF------- xenopus.tropicalis
 [5]  1062 MTLQNYYYYYYYYFFFFETKSCS...ILRLV--AGRIKNTF------- pongo.abelii
 ...   ... ...
 [8]  1062 -----------------------...ILVLV--AARIKNTF------- mus.musculus
 [9]  1062 -----------------------...ILVLV--AAKLKNTF------- gallus.gallus
[10]  1062 -----MAAAAASSSASGASSSSS...ILVLV--AAKIKNTF------- anolis.carolinensis
[11]  1062 -----------------------...IYII------------------ monodelphis.domes...
[12]  1062 -----------------------...ILVLV--AAKIKNAF------- ornithorhynchus.a...

$`108764at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]  1050 -----------------------...LANQTI--HKYCVV-------- ciona.savignyi
 [2]  1050 MSGFSSELIDYLEGRITFEEFDQ...MAHQLI--NTHCII-------- gasterosteus.acul...
 [3]  1050 ----MSELNDYLEGKITFEEFEK...MARHII--YTYCTI-------- danio.rerio
 [4]  1050 MSGFSPELIDYLEGKISFEEFER...MARRLL--HQYCVS-------- xenopus.tropicalis
 [5]  1050 MSGFSPELIDYLEGKISFEEFER...LARRLL--YTWGSV-------- ornithorhynchus.a...
 ...   ... ...
 [8]  1050 MSGFSPELIDYLEGKISFEEFER...MARKLL--YTYGTI-------- monodelphis.domes...
 [9]  1050 MSGFSPELIDYLEGKISFEEFER...MAQKLL--YTYCVI-------- mus.musculus
[10]  1050 MSGFSPELIDYLEGKIS----SS...MAQTLL--YTYCSI-------- pan.troglodytes
[11]  1050 MSGFSPELIDYLEGKISFEEFER...MAQTLL--YTYCSI-------- homo.sapiens
[12]  1050 MSGFSPELIDYLEGKISFEEFER...MAQRLL--YTYCSI-------- pongo.abelii</code></pre>
</div>
</div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="margin-aside">You can also use the function <code>msaPrettyPrint()</code> from the <em><a href="https://bioconductor.org/packages/3.19/ggmsa">ggmsa</a></em> package to visualize the MSA. This function will create a PDF in your working environment with a publication-ready figure (often with multiple pages, as MSAs are typically long) displaying your MSA</span></div></div>
<p>I guess you can see that some alignments have columns with a lot of gaps. Since these are usually not much informative for phylogeny inference, it is common practice to remove MSA columns with too many gaps. This process is called ‘trimming’. Here, we will the helper function <code>trim_alignment()</code> (source code in file <em>utils.R</em>) to remove alignment columns with &gt;50% of gaps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim MSAs with remove columns with &gt;50% of gaps</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>trimmed_msas <span class="ot">&lt;-</span> <span class="fu">lapply</span>(msas_aa, trim_alignment, <span class="at">max_gap =</span> <span class="fl">0.5</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at trimmed MSAs</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(trimmed_msas, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`102804at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]  1047 MSISLDKDAFFRRLKKIYSTWEN...-------------------RRR ciona.savignyi
 [2]  1047 -----------------------...WGCWG----------------- gallus.gallus
 [3]  1047 MAVTLDKEAYYRRIKRFYGNWKK...-----------PAPNPSKKRKK xenopus.tropicalis
 [4]  1047 MALSLDREAYYRRIKRLYSNWQK...SSGHSHSRSPGRSSAPPKKKRK anolis.carolinensis
 [5]  1047 MAVTLDKDAYYRRVKRLYSNWRK...SSGRGSNRGSRHSSAPPKKKRK homo.sapiens
 ...   ... ...
 [8]  1047 MAVTLDKDAYYRRVKRLYSNWRK...SSGRGSNRGSRHSSAPPKKKRK mus.musculus
 [9]  1047 MAVTLDKEAYYRRVKRLYSNWRK...SSGRGSSRGSRHSSAPPKKKRK monodelphis.domes...
[10]  1047 MAVTLDKEAYYRRVKRLYSNWRK...SSGRGSGRGSRHSSAPPKKKRK ornithorhynchus.a...
[11]  1047 MAVSLDKDAFYRRIKRLYGNWKK...-------RSSAPPPRSKKKRRH danio.rerio
[12]  1047 MAVNLDKEAYYRRIKRFYGNWKK...---------------------- gasterosteus.acul...

$`107151at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]   941 -----------------------...KNWKLKTKELLPIISSKVPVAF ciona.savignyi
 [2]   941 MGKKKKPFLGMPAPLGYVPGLGR...NTWRTWSCETVSRTSAVSRSVF gasterosteus.acul...
 [3]   941 LNKKKKPFLGMPAPLGYVPGLGR...LNWQKNIGEILVLVASKIKNSF danio.rerio
 [4]   941 MNKKKKPFLGMPAPLGYVPGLGR...KNWQQKIGEILVLVAAQIKNAF xenopus.tropicalis
 [5]   941 --QWRDLIIQTPGVL----GVSC...ANWQKKIGDILRLVAGRIKNTF pongo.abelii
 ...   ... ...
 [8]   941 MNKKKKPFLGMPAPLGYVPGLGR...TNWQRKIGEILVLVAARIKNTF mus.musculus
 [9]   941 MNKKKKPFLGMPAPLGYVPGLGR...VNWQKKIGEILVLVAAKLKNTF gallus.gallus
[10]   941 MNRKKKPFLGMPAPLGYVPGLGR...NNWQSKIGEILVLVAAKIKNTF anolis.carolinensis
[11]   941 MNKKKKPFLGMPAPLGYVPGLGR...---KEQVSNIYII--------- monodelphis.domes...
[12]   941 MNKKKKPFLGMPAPLGYVPGLGR...ENWQKKIGEILVLVAAKIKNAF ornithorhynchus.a...

$`108764at33208`
AAStringSet object of length 12:
     width seq                                              names               
 [1]   902 -----------------------...LIYRKSGNIDLANQTIHKYCVV ciona.savignyi
 [2]   902 MSGFSSELIDYLEGRITFEEFDQ...LIYQASGNMGMAHQLINTHCII gasterosteus.acul...
 [3]   902 ----MSELNDYLEGKITFEEFEK...LIYQSSGNKDMARHIIYTYCTI danio.rerio
 [4]   902 MSGFSPELIDYLEGKISFEEFER...LIYQNSGNLEMARRLLHQYCVS xenopus.tropicalis
 [5]   902 MSGFSPELIDYLEGKISFEEFER...LIYQNSGNLGLARRLLYTWGSV ornithorhynchus.a...
 ...   ... ...
 [8]   902 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNMGMARKLLYTYGTI monodelphis.domes...
 [9]   902 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNTAMAQKLLYTYCVI mus.musculus
[10]   902 MSGFSPELIDYLEGKIS----SS...LIYQSSGNTGMAQTLLYTYCSI pan.troglodytes
[11]   902 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNTGMAQTLLYTYCSI homo.sapiens
[12]   902 MSGFSPELIDYLEGKISFEEFER...LIYQSSGNTGMAQRLLYTYCSI pongo.abelii</code></pre>
</div>
</div>
<p>Now that we have trimmed MSAs, we can use them to infer a phylogeny.</p>
</section>
<section id="phylogeny-inference-concatenation-based-approaches" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="phylogeny-inference-concatenation-based-approaches"><span class="header-section-number">1.5</span> Phylogeny inference: concatenation-based approaches</h2>
<p>The simplest way of inferring a phylogeny from phylogenomic (i.e.&nbsp;multi-gene) data sets is to concatenate MSAs for each gene into a single MSA, and then infer a phylogeny as you’d do for a single MSA. This concatenated MSA is often referred to as a ‘supermatrix’, which is why this approach is usually called ‘supermatrix-based’, or ‘concatenation-based’.</p>
<p>To demonstrate how this works, we will concatenate MSAs and infer a species tree using IQ-TREE 2 <span class="citation" data-cites="minh2020iq">(<a href="#ref-minh2020iq" role="doc-biblioref">Minh et al. 2020</a>)</span>. For that, we will use a helper function <code>infer_species_tree()</code> (source code in <em>utils.R</em>), which runs IQ-TREE 2 from the R session and reads the inferred tree as an object of class <code>phylo</code> (standard class to represent trees in R, from the <em><a href="https://CRAN.R-project.org/package=ape">ape</a></em> package).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Concatenate alignments</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>conc_msa <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(xscat, trimmed_msas)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(conc_msa) <span class="ot">&lt;-</span> <span class="fu">names</span>(trimmed_msas[[<span class="dv">1</span>]])</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the concatenated alignment</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(conc_msa)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Infer species tree from concatenated alignment</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"ciona.savignyi"</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>tree_conc <span class="ot">&lt;-</span> <span class="fu">infer_species_tree</span>(</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">list</span>(<span class="at">conc =</span> conc_msa),</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    outgroup, </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The tree inferred from the concatenated alignment is stored in <code>output/trees/supermatrix.treefile</code>, and the output file with a detailed summary (models and AICs/BICs, base frequencies, etc) is stored in <code>output/trees/supermatrix.iqtree</code>. To load a phylogenetic tree from a file, you can use the function <code>read.tree()</code> from the R package <em><a href="https://CRAN.R-project.org/package=ape">ape</a></em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read tree as a `phylo` object</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>tree_conc <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"trees"</span>, <span class="st">"supermatrix.treefile"</span>))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot tree</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tree_conc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="phylogenomics_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="phylogeny-inference-coalescent-based-approaches" class="level2" data-number="1.6">
<h2 data-number="1.6" class="anchored" data-anchor-id="phylogeny-inference-coalescent-based-approaches"><span class="header-section-number">1.6</span> Phylogeny inference: coalescent-based approaches</h2>
<p>Here, instead of inferring a single tree from concatenated MSAs, we will infer an individual tree for each BUSCO gene. We will then use these set of trees to calculate concordance factors and use coalescent-based approaches.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Infer single-locus trees</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>outgroup <span class="ot">&lt;-</span> <span class="st">"ciona.savignyi"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>tree_sl <span class="ot">&lt;-</span> <span class="fu">infer_singlelocus_trees</span>(</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    trimmed_msas,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    outgroup,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">outdir =</span> <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>),</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">threads =</span> <span class="dv">4</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we will read the output tree file. Since this is not one, but many trees, trees will be stored in an object of class <code>multiPhylo</code>. In essence, a <code>multiPhylo</code> object is a list of multiple <code>phylo</code> objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read tree</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>tree_sl <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"sl.treefile"</span>))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>tree_sl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>282 phylogenetic trees</code></pre>
</div>
</div>
<p>Next, we will infer a species tree from this set of gene trees using ASTRAL <span class="citation" data-cites="zhang2017astral">(<a href="#ref-zhang2017astral" role="doc-biblioref">Zhang, Sayyari, and Mirarab 2017</a>)</span>, a Java program that can be run on the command line using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">java</span> <span class="at">-jar</span> astral.5.7.8.jar <span class="at">-i</span> sl.treefile <span class="at">-o</span> astral.tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we can read the species tree created by ASTRAL. Note that ASTRAL is a coalescent-based method that uses parsimony principles to infer <strong>a single species tree</strong> from <strong>a set of (many) gene trees</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read species tree created with ASTRAL</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>astral_tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"astral.tree"</span>))</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot tree</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(astral_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="phylogenomics_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Finally, once we have a representative species tree (generated by ASTRAL) and multiple gene trees (generated by IQ-TREE 2), we can calculate so-called <em>gene concordance factors (gCF)</em>, which represent the proportion of decisive gene trees that support each branch of the species tree. These can be computed with the code below</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate gCF with IQ-TREE 2</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"-t"</span>, <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"astral.tree"</span>),</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"--gcf"</span>, <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"sl.treefile"</span>),</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"--prefix"</span>, <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"gcf"</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(<span class="st">"iqtree2"</span>, args)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, we can read the tree and explore the tree with gCF for each branch with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read tree with gCF</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>gcf_tree <span class="ot">&lt;-</span> <span class="fu">read.tree</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"sl_trees"</span>, <span class="st">"gcf.cf.tree"</span>))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gcf_tree, <span class="at">show.node.label =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="phylogenomics_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="session-information" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-information">Session information</h2>
<p>This document was created under the following conditions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=nl_BE.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=nl_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=nl_BE.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=nl_BE.UTF-8 LC_IDENTIFICATION=C       

time zone: Arctic/Longyearbyen
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] msa_1.36.1          ape_5.8             lubridate_1.9.3    
 [4] forcats_1.0.0       stringr_1.5.1       dplyr_1.1.4        
 [7] purrr_1.0.2         readr_2.1.5         tidyr_1.3.1        
[10] tibble_3.2.1        ggplot2_3.5.1       tidyverse_2.0.0    
[13] Biostrings_2.72.1   GenomeInfoDb_1.40.1 XVector_0.44.0     
[16] IRanges_2.38.1      S4Vectors_0.42.1    BiocGenerics_0.50.0
[19] cogeqc_1.8.0        here_1.0.1         

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1        farver_2.1.2            vipor_0.4.7            
 [4] fastmap_1.2.0           lazyeval_0.2.2          promises_1.3.0         
 [7] digest_0.6.36           timechange_0.3.0        lifecycle_1.0.4        
[10] processx_3.8.4          tidytree_0.4.6          magrittr_2.0.3         
[13] compiler_4.4.1          rlang_1.1.4             tools_4.4.1            
[16] igraph_2.0.3            utf8_1.2.4              yaml_2.3.9             
[19] knitr_1.48              labeling_0.4.3          htmlwidgets_1.6.4      
[22] bit_4.0.5               curl_5.2.1              xml2_1.3.6             
[25] plyr_1.8.9              websocket_1.4.2         aplot_0.2.3            
[28] withr_3.0.0             grid_4.4.1              fansi_1.0.6            
[31] colorspace_2.1-0        scales_1.3.0            cli_3.6.3              
[34] rmarkdown_2.27          crayon_1.5.3            treeio_1.28.0          
[37] generics_0.1.3          rstudioapi_0.16.0       ggtree_3.12.0          
[40] httr_1.4.7              reshape2_1.4.4          tzdb_0.4.0             
[43] chromote_0.3.1          ggbeeswarm_0.7.2        cachem_1.1.0           
[46] zlibbioc_1.50.0         rvest_1.0.4             parallel_4.4.1         
[49] ggplotify_0.1.2         BiocManager_1.30.23     yulab.utils_0.1.4      
[52] vctrs_0.6.5             jsonlite_1.8.8          gridGraphics_0.5-1     
[55] hms_1.1.3               patchwork_1.2.0         bit64_4.0.5            
[58] beeswarm_0.4.0          glue_1.7.0              ps_1.7.7               
[61] stringi_1.8.4           gtable_0.3.5            later_1.3.2            
[64] UCSC.utils_1.0.0        munsell_0.5.1           pillar_1.9.0           
[67] htmltools_0.5.8.1       GenomeInfoDbData_1.2.12 R6_2.5.1               
[70] rprojroot_2.0.4         vroom_1.6.5             evaluate_0.24.0        
[73] lattice_0.22-6          BiocStyle_2.32.1        memoise_2.0.1          
[76] ggfun_0.1.5             Rcpp_1.0.13             nlme_3.1-165           
[79] xfun_0.46               fs_1.6.4                pkgconfig_2.0.3        </code></pre>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-harrison2024ensembl" class="csl-entry" role="listitem">
Harrison, Peter W, M Ridwan Amode, Olanrewaju Austine-Orimoloye, Andrey G Azov, Matthieu Barba, If Barnes, Arne Becker, et al. 2024. <span>“Ensembl 2024.”</span> <em>Nucleic Acids Research</em> 52 (D1): D891–99.
</div>
<div id="ref-minh2020iq" class="csl-entry" role="listitem">
Minh, Bui Quang, Heiko A Schmidt, Olga Chernomor, Dominik Schrempf, Michael D Woodhams, Arndt Von Haeseler, and Robert Lanfear. 2020. <span>“IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era.”</span> <em>Molecular Biology and Evolution</em> 37 (5): 1530–34.
</div>
<div id="ref-zhang2017astral" class="csl-entry" role="listitem">
Zhang, Chao, Erfan Sayyari, and Siavash Mirarab. 2017. <span>“ASTRAL-III: Increased Scalability and Impacts of Contracting Low Support Branches.”</span> In <em>RECOMB International Workshop on Comparative Genomics</em>, 53–75. Springer.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link  aria-label=" preface"="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>